cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractNotificationSubscriberService.java

```
@@ -0,0 +1,276 @@
+            } catch (JsonSyntaxException e) {
+              LOG.warn("Failed to decode message with id {}. Skipped. ", message.getId(), e);
+              // lastFetchedMessageId was updated at the beginning of the loop to skip this message in next fetch
+            }
+          }
+        }
+
+        // Sleep for configured number of milliseconds if there are no notifications
+        if (lastFetchedMessageId == null) {
+          return cConf.getLong(Constants.Scheduler.EVENT_POLL_DELAY_MILLIS);
+        }
+
+        final String lastMessageId = lastFetchedMessageId;
+        // Execute processed notifications in one transaction
+        Retries.supplyWithRetries(new Supplier<Void>() {
+          @Override
+          public Void get() {
+            Transactions.executeUnchecked(transactional, new TxCallable<Void>() {
+              @Override
+              public Void call(DatasetContext context) throws Exception {
+                processFetchedNotifications(context, fetchedNotifications);
+                updateMessageId(context, lastMessageId);
+                return null;
+              }
+            });
+            return null;
+          }
+        }, co.cask.cdap.common.service.RetryStrategies.fixDelay(Constants.Retry.RUN_RECORD_UPDATE_RETRY_DELAY_SECS,
+                                                                TimeUnit.SECONDS));
+
+        // Transaction was successful, so reset the failure count to 0
+        failureCount = 0;
+        // The last fetched message id was persisted successfully, update the local field as well.
+        messageId = lastMessageId;
+        return 0L;
+
+      } catch (ServiceUnavailableException e) {
+        SAMPLING_LOG.warn("Failed to contact service {}. Will retry in next run.", e.getServiceName(), e);
+      } catch (TopicNotFoundException e) {
+        SAMPLING_LOG.warn("Failed to fetch from TMS. Will retry in next run.", e);
+      } catch (Exception e) {
```

should we catch all exceptions?

---

I think it is okay - we need to retry fetching notifications anyway so I think it's better to log it and not continue throwing an error.